jscomp/syntax/src/res_core.ml

```diff
@@ -2815,21 +2822,23 @@ and parseJsxChildren p =
         let () = p.token <- token in
         children
     | token when Grammar.isJsxChildStart token ->
-      let () = Scanner.popMode p.scanner Jsx in
       let child =
         parsePrimaryExpr ~operand:(parseAtomicExpr p) ~noCall:true p
       in
       loop p (child :: children)
     | _ -> children
   in
-  match p.Parser.token with
-  | DotDotDot ->
-    Parser.next p;
-    (true, [parsePrimaryExpr ~operand:(parseAtomicExpr p) ~noCall:true p])
-  | _ ->
-    let children = List.rev (loop p []) in
-    Scanner.popMode p.scanner Jsx;
-    (false, children)
+  let spread, children =
+    match p.Parser.token with
+    | DotDotDot ->
+      Parser.next p;
+      (true, [parsePrimaryExpr ~operand:(parseAtomicExpr p) ~noCall:true p])
+    | _ ->
+      let children = List.rev (loop p []) in
+      (false, children)
+  in
+  Scanner.setJsxMode p.scanner;
```

Comment:
once we're done parsing the children, we reset JSX mode because we have to parse the closing tag.